repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/AbstractKeystoneV2Filter.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.filters.keystonev2
+
+import javax.servlet._
+import javax.servlet.http.HttpServletResponse._
+import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
+
+import com.rackspace.httpdelegation.HttpDelegationManager
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.openrepose.commons.utils.http.{IdentityStatus, OpenStackServiceHeader}
+import org.openrepose.commons.utils.servlet.http.ResponseMode.{MUTABLE, PASSTHROUGH}
+import org.openrepose.commons.utils.servlet.http.{HttpServletRequestWrapper, HttpServletResponseWrapper}
+import org.openrepose.core.filter.AbstractConfiguredFilter
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.filters.keystonev2.config._
+
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success, Try}
+
+abstract class AbstractKeystoneV2Filter[T <: KeystoneV2Config: ClassTag](configurationService: ConfigurationService)
+  extends AbstractConfiguredFilter[T](configurationService)
+    with HttpDelegationManager
+    with LazyLogging {
+
+  import AbstractKeystoneV2Filter._
+
+  def doAuth(request: HttpServletRequestWrapper): Try[Unit.type]
+  def handleFailures(authResult: Try[Unit.type]): Option[Reject]
+
+  def doWork(servletRequest: HttpServletRequest, servletResponse: HttpServletResponse, chain: FilterChain): Unit = {
+    /**
+      * STATIC REFERENCE TO CONFIG
+      */
+    val config = configuration
+
+    /**
+      * DECLARE COMMON VALUES
+      */
+    lazy val request = new HttpServletRequestWrapper(servletRequest)
+    lazy val response = new HttpServletResponseWrapper(servletResponse, MUTABLE, PASSTHROUGH)
+
+    def isWhitelisted(requestUri: String): Boolean = {
```

Comment:
You know, I'm not really sure why I nested these methods in the first place. I think I've come to regret that. Not that it's in scope of this story.

However, if we are going to leave these nested, I personally prefer them to be placed below the actual work being done by the containing method. To me, it makes it easier to see what this method is meant to be doing. Thought I admit it is strange to invoke a method before it has been defined.

Reply:
Oh. I guess I nested them so that I didn't have to keep passing the static `config` reference around.

Reply:
Weirdly it wouldn't let me put them below any more, irc. There were a couple of places where it things use to be defined below their usage and it didn't care, but now it does. I didn't bother trying to fight it i just made it happy.

Reply:
Oh, huh, weird. I was going to say, I know a bunch of languages either don't support calling methods/functions before they are defined, or they make you forward declare them, but I thought it would "just work" with Scala (mostly because it seemingly does work with some of the other nested things).

Oh well.